sdk/tests/AccountIdTest.cc

```diff
@@ -0,0 +1,128 @@
+}
+
+TEST_F(AccountIdTest, ConstructFromString)
+{
+  const std::string testShardNumStr = std::to_string(getTestShardNum());
+  const std::string testRealmNumStr = std::to_string(getTestRealmNum());
+  const std::string testAccountNumStr = std::to_string(getTestAccountNum());
+
+  AccountId accountId(testShardNumStr + '.' + testRealmNumStr + '.' + testAccountNumStr);
+  EXPECT_EQ(accountId.getShardNum(), getTestShardNum());
+  EXPECT_EQ(accountId.getRealmNum(), getTestRealmNum());
+  EXPECT_EQ(accountId.getAccountNum(), getTestAccountNum());
+
+  EXPECT_THROW(AccountId(".." + testShardNumStr + testRealmNumStr + testAccountNumStr), std::invalid_argument);
+  EXPECT_THROW(AccountId('.' + testShardNumStr + '.' + testRealmNumStr + testAccountNumStr), std::invalid_argument);
+  EXPECT_THROW(AccountId('.' + testShardNumStr + testRealmNumStr + '.' + testAccountNumStr), std::invalid_argument);
+  EXPECT_THROW(AccountId('.' + testShardNumStr + testRealmNumStr + testAccountNumStr + '.'), std::invalid_argument);
+  EXPECT_THROW(AccountId(testShardNumStr + ".." + testRealmNumStr + testAccountNumStr), std::invalid_argument);
+  EXPECT_THROW(AccountId(testShardNumStr + '.' + testRealmNumStr + testAccountNumStr + '.'), std::invalid_argument);
+  EXPECT_THROW(AccountId(testShardNumStr + testRealmNumStr + ".." + testAccountNumStr), std::invalid_argument);
+  EXPECT_THROW(AccountId(testShardNumStr + testRealmNumStr + '.' + testAccountNumStr + '.'), std::invalid_argument);
+  EXPECT_THROW(AccountId('.' + testShardNumStr + '.' + testRealmNumStr + '.' + testAccountNumStr + '.'),
+               std::invalid_argument);
+
+  EXPECT_THROW(AccountId("abc"), std::invalid_argument);
+  EXPECT_THROW(AccountId("o.o.e"), std::invalid_argument);
+  EXPECT_THROW(AccountId("0.0.1!"), std::invalid_argument);
+}
+
+TEST_F(AccountIdTest, SetShardRealmaccountnum)
+{
+  AccountId accountId;
+  accountId.setShardNum(getTestShardNum());
+  accountId.setRealmNum(getTestRealmNum());
+  accountId.setAccountNum(getTestAccountNum());
+
+  EXPECT_EQ(accountId.getShardNum(), getTestShardNum());
+  EXPECT_EQ(accountId.getRealmNum(), getTestRealmNum());
+  EXPECT_EQ(accountId.getAccountNum(), getTestAccountNum());
+}
+
+TEST_F(AccountIdTest, ProtobufaccountId)
+{
+  AccountId accountId;
+  accountId.setShardNum(getTestShardNum());
+  accountId.setRealmNum(getTestRealmNum());
+  accountId.setAccountNum(getTestAccountNum());
+
+  auto protoAccountId = std::unique_ptr<proto::AccountID>(accountId.toProtobuf());
+  EXPECT_EQ(protoAccountId->shardnum(), getTestShardNum());
+  EXPECT_EQ(protoAccountId->realmnum(), getTestRealmNum());
+  EXPECT_EQ(protoAccountId->accountnum(), getTestAccountNum());
+
+  const uint64_t adjustment = 3ULL;
+  const uint64_t newShard = getTestShardNum() + adjustment;
+  const uint64_t newRealm = getTestRealmNum() - adjustment;
+  const uint64_t newContract = getTestAccountNum() * adjustment;
+
+  protoAccountId->set_shardnum(static_cast<int64_t>(newShard));
```

Comment:
Seeing this test case brings to mind - 

why are `AccountId` member variables `uint_64`, if the protobuf fields are `int64_t`? As it stands, there are a bunch of `uint_64` values which could be successfully set on the class, that would produce undefined behavior with this static cast

Reply:
For some reason the protobuf fields aren't generated as `uint64_t` even though they should be nonnegative. I'm not sure if this is a limitation of protobuf or what, but I felt it should be enforced in the SDK that these numbers can't be negative. Maybe I can just put a comment for the class to saying that 2^64 - 1 is the max value for each of these fields?

Reply:
If the protobuf has `int64`, I would say the underlying object should as well. Non-negative values could be enforced by `if`s in the constructors and setters.

IMO a comment isn't enough, since they can be ignored. If the underlying class values are kept as `uint64`, there would need to be an `if` block anyway, to make sure the value isn't too big. Either way, a manual check is necessary

Reply:
I'll add some throws for input numbers that are too big.

I personally like keeping the `uint64_t` type just to enforce in the code that the type should be nonnegative. The way I see it, either way `if` blocks will have to be added to verify values are within range (either `int64_t` is nonnegative or `uint64_t` isn't too big) and comments will be added to specify ranges to the user. The `uint64_t` type I feel better exemplifies how the value should be seen and used. With that said my opinions not super strong so I'm willing to change it if it causes heartburn.

Reply:
I'm fine with this way :+1:

A couple things remaining-
1. the string constructor also needs the max value checks
2. I would personally add some tests to make sure the checks throughout `AccountId` are hit, passing in `int64 max` and `int64 max +1` to verify. Up to you, though

Reply:
Tests added